compiler/rustc_expand/src/mbe/macro_rules.rs

```diff
@@ -542,8 +542,17 @@ pub fn compile_declarative_macro(
     // Compute the spans of the macro rules
     // We only take the span of the lhs here,
     // so that the spans of created warnings are smaller.
+    // Also, we are only interested in non-foreign macros.
     let rule_spans = if def.id != DUMMY_NODE_ID {
```

Comment:
```suggestion
    let rule_spans = if valid && def.id != DUMMY_NODE_ID {
```
This is off-topic, but we don't need to mark the arms as unused for malformed macros, this would only result in additional diagnostic noise.

Reply:
Oh good point currently this:

```rust
#![deny(unused_macro_rules)]

macro_rules! foo {
    (v) => {};
    () => 0;
}
fn main() {
    foo!(v);
}
```

gives:

```
error: macro rhs must be delimited
 --> src/main.rs:5:11
  |
5 |     () => 0;
  |           ^

error: 1st rule of macro `foo` is never used
 --> src/main.rs:4:5
  |
4 |     (v) => {};
  |     ^^^
  |
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![deny(unused_macro_rules)]
  |         ^^^^^^^^^^^^^^^^^^

error: 2nd rule of macro `foo` is never used
 --> src/main.rs:5:5
  |
5 |     () => 0;
  |     ^^

```

which is very wrong, because the first arm is very much used. For some reason, the unused macros lint does not have this issue, as in this does lint:

```Rust
#![deny(unused_macro_rules)]

macro_rules! foo {
    (v) => {};
    () => 0;
}

fn main() {}
```
```
error: macro rhs must be delimited
 --> src/main.rs:5:11
  |
5 |     () => 0;
  |           ^

warning: unused macro definition: `foo`
 --> src/main.rs:3:14
  |
3 | macro_rules! foo {
  |              ^^^
  |
  = note: `#[warn(unused_macros)]` on by default
```

But the unused_macros lint is gone if you use `foo`. I guess it might be an idea to add a test for it.